src/engine/OrderedCollection.hpp

```diff
@@ -0,0 +1,139 @@
+#pragma once
+#include <iterator>
+#include <map>
+#include <memory>
+#include <set>
+#include <vector>
+
+#include "IOrdered.hpp"
+
+namespace admirals {
+
+/// @brief Maps string keys (`IOrdered::name`) to `IOrdered` object pointers and
+/// uses a separate collection to order the contained `IOrdered` object pointers
+/// by their `IOrdered::order` value. The objects are ordered from low to hight.
+/// Using these two collections, it can quickly search among objects in the
+/// collection using their `IOrdered::name` value, while also keeping the
+/// objects sorted by their `IOrdered::order` value`.
+/// @tparam T The contained class implementing the `IOrdered` interface.
+template <typename T = IOrdered> class OrderedCollection {
+private:
+    /// @brief Compares values of the multiset to order the values by their
+    //         respective `IOrdered::order` value.
+    struct Comparator {
+        Comparator(const OrderedCollection &c) : m_collection(c) {}
+
+        bool operator()(const std::string &l, const std::string &r) const {
+            const std::shared_ptr<IOrdered> a = m_collection.m_objects.at(l);
+            const std::shared_ptr<IOrdered> b = m_collection.m_objects.at(r);
+            return a->order() < b->order();
+        }
+
+    private:
+        const OrderedCollection &m_collection;
+    };
+
+    typedef std::map<std::string, std::shared_ptr<T>> KeyToPointerMap;
+    typedef std::multiset<std::string, Comparator> OrderedKeySet;
+
+public:
+    OrderedCollection() : m_ordered(OrderedKeySet(Comparator(*this))) {}
+
+    /// @brief Iterates over the `IOrdered` object points in the collection in
+    /// the same order as the `IOrdered::name` values are sorted in the ordered
+    /// multiset.
+    struct Iterator {
+        using iterator_category = std::forward_iterator_tag;
+        using difference_type = std::ptrdiff_t;
+        using value_type = const std::shared_ptr<T>;
+        using pointer = const std::shared_ptr<T> *;
+        using reference = const std::shared_ptr<T> &;
+
+        Iterator(OrderedKeySet::iterator i, const KeyToPointerMap &m_objects)
+            : m_iter(i), m_objects(m_objects) {}
+
+        reference operator*() const {
+            auto v = (m_iter.operator*)();
+            return m_objects.at(v);
+        }
+
+        pointer operator->() {
+            auto v = (m_iter.operator*)();
+            return &m_objects.at(v);
+        }
```

Comment:
The semantics here become a bit fuzzy. We want the iterator to iterate over the actual elements in the map (i.e `std::share_ptr<IOrdered>` in the end), but we do not want to return a pointer/reference to the `std::shared_ptr<IOrdered>` to the user, we want to return a copy of the value.

I also feel that this isn't really a problem in our use case, but everything is very implicit. If we return a copy, everything becomes safely handled.